lib/tesla/middleware/consul_watch.ex

```diff
@@ -14,69 +14,92 @@ defmodule Tesla.Middleware.ConsulWatch do
-    GenServer.start_link(__MODULE__, [], name: __MODULE__)
-  end
-
-  @impl GenServer
-  def init(_) do
-    @index_table = :ets.new(@index_table, [:named_table, :public])
+  @impl Tesla.Middleware
+  def call(%{url: url} = env, next, opts) do
+    wait = Keyword.get(opts, :wait)
 
-    {:ok, []}
+    env
+    |> maybe_set_index(url)
+    |> maybe_set_wait(wait)
+    |> Tesla.run(next)
+    |> store_index(url)
   end
 
-  @impl GenServer
-  def handle_call({:reset, url}, _from, state) do
-    :ets.delete(@index_table, url)
+  defp maybe_set_index(%{method: :get, query: query} = env, url) do
+    index =
+      case Keyword.fetch(query, :index) do
+        {:ok, index} ->
+          index
 
-    {:reply, :ok, state}
-  end
+        _ ->
+          IndexStore.get_index(url)
+      end
 
-  @impl Tesla.Middleware
-  def call(env, next, opts) do
-    env
-    |> load_index(opts)
-    |> Tesla.run(next)
-    |> store_index()
+    %{env | query: Keyword.put(query, :index, index)}
   end
 
-  defp load_index(%{method: :get, url: url} = env, opts) do
-    case current_index(url) do
+  defp maybe_set_wait(%{query: query} = env, wait) do
+    case Keyword.get(query, :index) do
       nil ->
+        # don't set wait param on initial fetch
         env
 
-      index ->
-        query =
-          case Keyword.get(opts, :wait) do
-            nil -> []
-            wait -> [wait: to_gotime(wait)]
-          end
+      _ ->
+        wait = Keyword.get(query, :wait, wait)
+        %{env | query: Keyword.put(query, :wait, format_wait(wait))}
+    end
+  end
 
-        query = [index: index] ++ query
+  defp store_index({:ok, env}, url) do
+    index =
+      with [index | _] <- Tesla.get_headers(env, @header_x_consul_index),
+           {index, _} <- Integer.parse(index) do
+        index
+      else
+        _ -> nil
+      end
 
-        Map.update!(env, :query, &Keyword.merge(&1, query))
-    end
+    handle_new_index(url, index)
+
+    {:ok, env}
   end
 
-  defp load_index(env, _opts), do: env
+  def store_index({:error, reason}), do: {:error, reason}
+
+  defp handle_new_index(url, nil) do
+    IndexStore.reset_index(url)
+  end
+
+  defp handle_new_index(url, new_index) do
+    current_index = IndexStore.get_index(url)
+
+    cond do
+      is_nil(current_index) ->
+        IndexStore.store_index(url, new_index)
+
+      new_index < current_index || new_index < 1 ->
+        # reset index when it has moved backwards or is not greater than 0
+        # see: https://www.consul.io/api-docs/features/blocking#implementation-details
+        IndexStore.reset_index(url)
 
-  defp current_index(url) do
-    case :ets.lookup(@index_table, url) do
-      [] -> nil
-      [{_, index}] -> index
+      true ->
+        IndexStore.store_index(url, new_index)
     end
   end
 
-  def to_gotime(duration) do
+  defp format_wait(nil), do: nil
+  defp format_wait(true), do: nil
+  defp format_wait(wait) when is_binary(wait), do: wait
+
+  defp format_wait(duration) do
```

Comment:
if someone passes `wait` as `false` it will crash right?
```suggestion
  defp format_wait(duration), when is_number(suggestion) do
```
And a fallback one at the end.

one more question on this:
- if the wait is nil it will take the `5mins` default duration?

Reply:
> if someone passes wait as false it will crash right?

hmm yes, honestly I'm not sure what would be expected behaviour, I didn't think about this case at all. 

> if the wait is nil it will take the 5mins default duration?

Yes, it will send the param as `?wait=` for which Consul will use its default (5m).

Reply:
@Zaherhs I reordered the clauses as per your suggestion. 

Now `wait: false` is going to fall into default clause too, but it's best to think of it as unspecified behaviour. It's not documented.